crates/handlers/src/views/recovery/progress.rs

```diff
@@ -62,3 +70,51 @@ pub(crate) async fn get(
 
     Ok((cookie_jar, Html(rendered)).into_response())
 }
+
+pub(crate) async fn post(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    mut repo: BoxRepository,
+    State(templates): State<Templates>,
+    State(url_builder): State<UrlBuilder>,
+    PreferredLanguage(locale): PreferredLanguage,
+    cookie_jar: CookieJar,
+    Path(id): Path<Ulid>,
+    Form(form): Form<ProtectedForm<()>>,
+) -> Result<Response, FancyError> {
+    let (session_info, cookie_jar) = cookie_jar.session_info();
+    let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+
+    let maybe_session = session_info.load_session(&mut repo).await?;
+    if maybe_session.is_some() {
+        // TODO: redirect to continue whatever action was going on
+        return Ok((cookie_jar, url_builder.redirect(&mas_router::Index)).into_response());
+    }
+
+    let Some(recovery_session) = repo.user_recovery().lookup_session(id).await? else {
+        // XXX: is that the right thing to do?
+        return Ok((
+            cookie_jar,
+            url_builder.redirect(&mas_router::AccountRecoveryStart),
+        )
+            .into_response());
```

Comment:
maybe we should say the link 'expired' or something. I guess.

Reply:
I mean it's really that you got to the wrong place, as `lookup_session` will give you even consumed/expired sessions. A 404 would be better, but we don't have a nice way to display that page.

I think we should track in a separate issue to have nice 404 pages for when you go to pages with an unknown ID, because we have many cases like that

Reply:
https://github.com/matrix-org/matrix-authentication-service/issues/2880

Reply:
well this kind of raises another Q to me: Do we really want to keep old expired sessions around in the DB forever?

I can't say I see much reason to keep them for a very long time, because it's just going to be ever-growing cruft.

Reply:
Generally we don't cleanup most things in the database. I rather soft-delete stuff and keep things for now, and introduce maintenance/cleanup tasks later once we have a nice way of doing this?

Reply:
cynical take: that's how you get Synapse ;-)